src/election.ts

```
@@ -0,0 +1,167 @@
+ */
+export class Election {
+  public static readonly prefix = 'election';
+  public static readonly notLeaderError = new Error('election: not leader');
+  public static readonly noLeaderError = new Error('election: no leader');
+  public static readonly notReadyError = new Error('election: no ready');
+
+  public readonly namespace: Namespace;
+  public readonly lease: Lease;
+
+  private _leaseId = '';
+  private _leaderKey = '';
+  private _leaderRevision = '';
+  private _isCampaigning = false;
+
+  public get leaseId(): string { return this._leaseId; }
+  public get leaderKey(): string { return this._leaderKey; }
+  public get leaderRevision(): string { return this._leaderRevision; }
+  public get isReady(): boolean { return this._leaseId.length > 0; }
+  public get isCampaigning(): boolean { return this._isCampaigning; }
+
+  constructor(namespace: Namespace,
+              public readonly name: string,
+              public readonly ttl: number = 60) {
+    this.namespace = namespace.namespace(this.getPrefix());
+    this.lease = this.namespace.lease(ttl);
+  }
+
+  public async ready() {
+    const leaseId = await this.lease.grant();
+
+    if (!this.isReady) {
+      this._leaseId = leaseId;
+    }
+  }
+
+  public async campaign(value: string) {
+    this.throwIfNotReady();
+    const result = await this.namespace
+      .if(this.leaseId, 'Create', '==', 0)
+      .then(this.namespace.put(this.leaseId).value(value).lease(this.leaseId))
+      .else(this.namespace.get(this.leaseId))
+      .commit();
+
+    this._leaderKey = `${this.getPrefix()}${this.leaseId}`;
+    this._leaderRevision = result.header.revision;
+    this._isCampaigning = true;
+
+    if (!result.succeeded) {
+      try {
+        const kv = result.responses[0].response_range.kvs[0];
+        this._leaderRevision = kv.create_revision;
+        if (kv.value.toString() !== value) {
+          await this.proclaim(value);
+        }
+      } catch (error) {
+        await this.resign();
```

We should catch and silence any errors that happen here (as the Go implementation does) which would be safe if you use the lease invalidation strategy described below.

---

It there any good ways let us tell user this campaign call has occurred error and should not keep going without throw error? Or we just return a new campaign call? I have no idea what case could lead to fail at this moment.

---

Maybe another way is to use a 'elected' and 'resigned' events.